drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,514 @@
+	} else {
+		irq_disable(config->irq);
+	}
+
+	return 0;
+}
+
+static int kscan_npcx_drive_column(const struct device *dev, int col)
+{
+	const struct kscan_npcx_config *config = dev->config;
+	struct kbs_reg *const inst = config->base;
+	uint32_t mask;
+
+	if (col >= CONFIG_KSCAN_NPCX_COLUMN_SIZE) {
+		return -EINVAL;
+	}
+
+	if (col == KEYBOARD_COLUMN_DRIVE_NONE) {
+		/* Drive all lines to high. ie. Key detection is disabled. */
+		mask = ~0;
+	} else if (col == KEYBOARD_COLUMN_DRIVE_ALL) {
+		/* Drive all lines to low for detection any key press */
+		mask = ~(BIT(CONFIG_KSCAN_NPCX_COLUMN_SIZE) - 1);
+	} else {
+		/*
+		 * Drive one line to low for determining which key's state
+		 * changed.
+		 */
+		mask = ~BIT(col);
+	}
+
+	LOG_DBG("%x", mask);
+
+	/* Set KBSOUT */
+	inst->KBSOUT0 = (mask & 0xFFFF);
+	inst->KBSOUT1 = ((mask >> 16) & 0x03);
+
+	return 0;
+}
+
+static int kscan_npcx_read_row(const struct device *dev, int *row)
+{
+	const struct kscan_npcx_config *config = dev->config;
+	struct kbs_reg *const inst = config->base;
+	int val;
+
+	val = inst->KBSIN;
+
+	/* 1 means key pressed, otherwise means key released. */
+	*row = (~val & NPCX_KB_ROW_MASK);
+
+	return 0;
+}
+
+static bool is_matrix_ghosting(const uint8_t *state)
+{
+	/*
+	 * matrix keyboard designs are suceptible to ghosting.
+	 * An extra key appears to be pressed when 3 keys
+	 * belonging to the same block are pressed.
+	 * for example, in the following block
+	 *
+	 * . . w . q .
+	 * . . . . . .
+	 * . . . . . .
+	 * . . m . a .
+	 *
+	 * the key m would look as pressed if the user pressed keys
+	 * w, q and a simultaneously. A block can also be formed,
+	 * with not adjacent columns.
+	 */
+	for (int c = 0; c < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c++) {
+		if (!state[c]) {
+			continue;
+		}
+
+		for (int c_next = c + 1; c_next < CONFIG_KSCAN_NPCX_COLUMN_SIZE; c_next++) {
+			/*
+			 * We AND the columns to detect a "block".
+			 * This is an indication of ghosting, due to current
+			 * flowing from a key which was never pressed. In our
+			 * case, current flowing is a bit set to 1 as we
+			 * flipped the bits when the matrix was scanned.
+			 * Now we OR the colums using z&(z-1) which is
+			 * non-zero only if z has more than one bit set.
+			 */
+			uint8_t common_row_bits = state[c] & state[c_next];
+
+			if (common_row_bits & (common_row_bits - 1)) {
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+static bool read_keyboard_matrix(const struct device *dev, uint8_t *new_state)
+{
+	int row;
+	uint8_t key_event = 0U;
+
+	for (int col = 0; col < CONFIG_KSCAN_NPCX_COLUMN_SIZE; col++) {
+		kscan_npcx_drive_column(dev, col);
+
+		/* Allow the matrix to stabilize before reading it */
+		k_busy_wait(CONFIG_KSCAN_NPCX_POLL_COL_OUTPUT_SETTLE_TIME);
+
+		kscan_npcx_read_row(dev, &row);
+		new_state[col] = row & 0xFF;
+		key_event |= row;
+	}
+
+	kscan_npcx_drive_column(dev, KEYBOARD_COLUMN_DRIVE_NONE);
+
+	return key_event != 0U ? true : false;
```

Comment:
Just use `return key_event != 0U`, or I think just `return key_event` should do exactly the same. If you really want to be explicit I would expand it but I don't see how this could be ambiguous.

Reply:
I's prefer to use `return key_event != 0U` because the return type of this function is bool. 
The data type of key_event is uint8_t.